src/main/java/com/solopov/urlshorter/sevices/impl/UrlShorterServiceImpl.java

```diff
@@ -0,0 +1,49 @@
+package com.solopov.urlshorter.sevices.impl;
+
+import com.solopov.urlshorter.dao.entities.ShortenedUrl;
+import com.solopov.urlshorter.dao.repositories.ShortenedUrlRepository;
+import com.solopov.urlshorter.exceptions.LinkNotFoundException;
+import com.solopov.urlshorter.sevices.UrlShorterService;
+import com.solopov.urlshorter.web.dto.UrlGenerateRequestDto;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.data.jpa.repository.Lock;
+import org.springframework.retry.annotation.Retryable;
+import org.springframework.stereotype.Service;
+
+import javax.persistence.LockModeType;
+
+@Service
+@RequiredArgsConstructor
+public class UrlShorterServiceImpl implements UrlShorterService {
+
+    private final ShortenedUrlRepository shortenedRepository;
+
+    @Value("${com.solopov.urlshorter.service.url}")
+    private String rootUrl;
+
+    @Value("${com.solopov.urlshorter.service.api.shorten.url}")
+    private String shortenUrlPath;
+
+
+    @Retryable
+    @Override
+    @Lock(LockModeType.WRITE)
```

Comment:
I don't know much about @Lock but the java doc says it's used with `@Query` which is usually applied to repositories, not at the service layer.

`@Transactional` ?

Reply:
upd commit: removed, because @Lock needs to investigate 
@Transactional also didn't add because we have only one save operation related to db